### PR TITLE
KEYCLOAK-12231 Add description to all the fields

### PR DESCRIPTION
**Pull request by slaskawi**
## JIRA ID

[KEYCLOAK-12231](https://issues.jboss.org/browse/KEYCLOAK-12231)
## Additional Information

This Pull Request adds a lot of documentation to the Go objects and CRDs. Note, that once comments are added to the code, we need to regenerate both Open API and K8s objects by invoking:

```
operator-sdk-local generate k8s
operator-sdk-local generate openapi
```

## Checklist:
- [ ] Verified by team member
- [ ] Comments where necessary
- [ ] Automated Tests
- [ ] Documentation changes if necessary

### CODE REVIEW

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go [hunk 1]
```diff
@@ -4,26 +4,60 @@ import (
...
+	// Using an external database requires providing a secret containing credentials
+	// as well as connection details. Here's an example of such secret:
+	//
+	//     apiVersion: v1
+	//     kind: Secret
+	//     metadata:
+	//         name: keycloak-db-secret
+	//         namespace: keycloak
+	//     data:
+	//         POSTGRES_DATABASE: cm9vdA==
```
Comment (stianst) [inline pkg/apis/keycloak/v1alpha1/keycloak_types.go]: Would be nicer if example used string data so it's readable

Comment (stianst) [reply pkg/apis/keycloak/v1alpha1/keycloak_types.go]: POSTGRES_EXTERNAL_PORT is this optional? If so perhaps mention it

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go [hunk 2]
```diff
@@ -4,26 +4,60 @@ import (
...
+	//         POSTGRES_PASSWORD: ZXhhbXBsZS1rZXljbG9hay1lTWFQ
+	//         POSTGRES_SUPERUSER: dHJ1ZQ==
+	//         POSTGRES_USERNAME: ZXhhbXBsZS1rZXljbG9hay1WbGdU
+	//      type: Opaque
+	//
+	// Both POSTGRES_EXTERNAL_ADDRESS and POSTGRES_EXTERNAL_PORT are specifically required for creating
+	// connection to the external database. The secret name is created using the following convention:
+	//       <Custom Resource Name>-db-secret
+	//
+	// For more information, please refer to the manual pages.
```
Comment (stianst) [inline pkg/apis/keycloak/v1alpha1/keycloak_types.go]: "Operator documentation"?

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go [hunk 3]
```diff
@@ -7,16 +7,53 @@ import (
 // KeycloakBackupSpec defines the desired state of KeycloakBackup
 // +k8s:openapi-gen=true
 type KeycloakBackupSpec struct {
-	Restore bool            `json:"restore,omitempty"`
-	AWS     KeycloakAWSSpec `json:"aws,omitempty"`
+	// Controls automatic restore behavior.
+	// Currently not implemented (yet)
```
Comment (stianst) [inline pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Remove (yet). Perhaps add a short paragraph on how it is envisioned to work in the future.

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go [hunk 4]
```diff
@@ -7,16 +7,53 @@ import (
...
+	// +optional
+	Restore bool `json:"restore,omitempty"`
+	// If provided, an automatic database backup will be created on AWS S3.
```
Comment (stianst) [inline pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Change wording to say backup will be done on AWS S3.

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go [hunk 5]
```diff
@@ -7,16 +7,53 @@ import (
...
+	//     kind: Secret
+	//     metadata:
+	//       name: <Secret name>
+	//     type: Opaque
+	//     stringData:
+	//       GPG_PUBLIC_KEY: <GPG Public Key>
+	//       GPG_TRUST_MODEL: <GPG Trust Model>
+	//       GPG_RECIPIENT: <GPG Recipient>
+	//
+	// For more information, please refer to the manual pages.
```
Comment (stianst) [inline pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: documentation

Comment (stianst) [reply pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Add that if specified database backup is encrypted, if not specified it's not

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go (hunk 1 above)
Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloak_types.go]: I've just checked - it's mandatory. Created [KEYCLOAK-12304](https://issues.jboss.org/browse/KEYCLOAK-12304).

Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloak_types.go]: Fixed.

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go (hunk 2 above)
Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloak_types.go]: Fixed.

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go (hunk 3 above)
Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Fixed.

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go (hunk 4 above)
Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Fixed.

File: pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go (hunk 5 above)
Comment (slaskawi) [reply pkg/apis/keycloak/v1alpha1/keycloakbackup_types.go]: Fixed.

**Comment by slaskawi** [comment]
@stianst Updated. Ready for the next round.

**Comment by abstractj** [comment]
@slaskawi [yamlint](https://github.com/adrienverge/yamllint) reported some issues that might worth to take a look:

```
./keycloak.org_keycloakrealms_crd.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (86 > 80 characters)  (line-length)
  21:81     error    line too long (81 > 80 characters)  (line-length)
  22:81     error    line too long (152 > 80 characters)  (line-length)
  25:81     error    line too long (82 > 80 characters)  (line-length)
  26:81     error    line too long (82 > 80 characters)  (line-length)
  27:81     error    line too long (156 > 80 characters)  (line-length)
  32:81     error    line too long (83 > 80 characters)  (line-length)
  38:81     error    line too long (89 > 80 characters)  (line-length)
  41:81     error    line too long (89 > 80 characters)  (line-length)
  42:81     error    line too long (85 > 80 characters)  (line-length)
  45:81     error    line too long (83 > 80 characters)  (line-length)
  49:81     error    line too long (82 > 80 characters)  (line-length)
  50:81     error    line too long (82 > 80 characters)  (line-length)
  54:81     error    line too long (89 > 80 characters)  (line-length)
  55:81     error    line too long (84 > 80 characters)  (line-length)
  56:81     error    line too long (83 > 80 characters)  (line-length)
  57:81     error    line too long (83 > 80 characters)  (line-length)
  63:21     error    wrong indentation: expected 22 but found 20  (indentation)
  70:81     error    line too long (82 > 80 characters)  (line-length)
  71:81     error    line too long (82 > 80 characters)  (line-length)
  72:81     error    line too long (82 > 80 characters)  (line-length)
  73:81     error    line too long (86 > 80 characters)  (line-length)
  101:81    error    line too long (82 > 80 characters)  (line-length)
  127:81    error    line too long (87 > 80 characters)  (line-length)
  133:81    error    line too long (90 > 80 characters)  (line-length)
  163:81    error    line too long (85 > 80 characters)  (line-length)
  191:81    error    line too long (83 > 80 characters)  (line-length)
  218:21    error    wrong indentation: expected 22 but found 20  (indentation)
  239:81    error    line too long (85 > 80 characters)  (line-length)
  257:81    error    line too long (85 > 80 characters)  (line-length)
  266:81    error    line too long (84 > 80 characters)  (line-length)
  298:81    error    line too long (87 > 80 characters)  (line-length)
  362:15    error    wrong indentation: expected 16 but found 14  (indentation)
  377:11    error    wrong indentation: expected 12 but found 10  (indentation)
  380:81    error    line too long (86 > 80 characters)  (line-length)
  386:81    error    line too long (82 > 80 characters)  (line-length)
  393:81    error    line too long (82 > 80 characters)  (line-length)
  401:81    error    line too long (88 > 80 characters)  (line-length)
  402:81    error    line too long (85 > 80 characters)  (line-length)
  406:11    error    wrong indentation: expected 12 but found 10  (indentation)
  414:3     error    wrong indentation: expected 4 but found 2  (indentation)

./keycloak.org_keycloakusers_crd.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (86 > 80 characters)  (line-length)
  21:81     error    line too long (81 > 80 characters)  (line-length)
  22:81     error    line too long (152 > 80 characters)  (line-length)
  25:81     error    line too long (82 > 80 characters)  (line-length)
  26:81     error    line too long (82 > 80 characters)  (line-length)
  27:81     error    line too long (156 > 80 characters)  (line-length)
  32:81     error    line too long (81 > 80 characters)  (line-length)
  35:81     error    line too long (82 > 80 characters)  (line-length)
  38:81     error    line too long (89 > 80 characters)  (line-length)
  41:81     error    line too long (89 > 80 characters)  (line-length)
  42:81     error    line too long (85 > 80 characters)  (line-length)
  45:81     error    line too long (83 > 80 characters)  (line-length)
  49:81     error    line too long (82 > 80 characters)  (line-length)
  50:81     error    line too long (82 > 80 characters)  (line-length)
  54:81     error    line too long (89 > 80 characters)  (line-length)
  55:81     error    line too long (84 > 80 characters)  (line-length)
  56:81     error    line too long (83 > 80 characters)  (line-length)
  57:81     error    line too long (83 > 80 characters)  (line-length)
  63:21     error    wrong indentation: expected 22 but found 20  (indentation)
  70:81     error    line too long (82 > 80 characters)  (line-length)
  71:81     error    line too long (82 > 80 characters)  (line-length)
  72:81     error    line too long (82 > 80 characters)  (line-length)
  73:81     error    line too long (86 > 80 characters)  (line-length)
  92:81     error    line too long (81 > 80 characters)  (line-length)
  155:11    error    wrong indentation: expected 12 but found 10  (indentation)
  158:81    error    line too long (84 > 80 characters)  (line-length)
  161:81    error    line too long (82 > 80 characters)  (line-length)
  168:11    error    wrong indentation: expected 12 but found 10  (indentation)
  174:3     error    wrong indentation: expected 4 but found 2  (indentation)

./keycloak.org_keycloakbackups_crd.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (86 > 80 characters)  (line-length)
  21:81     error    line too long (81 > 80 characters)  (line-length)
  22:81     error    line too long (152 > 80 characters)  (line-length)
  25:81     error    line too long (82 > 80 characters)  (line-length)
  26:81     error    line too long (82 > 80 characters)  (line-length)
  27:81     error    line too long (156 > 80 characters)  (line-length)
  32:81     error    line too long (85 > 80 characters)  (line-length)
  35:81     error    line too long (84 > 80 characters)  (line-length)
  36:81     error    line too long (83 > 80 characters)  (line-length)
  40:81     error    line too long (81 > 80 characters)  (line-length)
  41:81     error    line too long (96 > 80 characters)  (line-length)
  42:81     error    line too long (89 > 80 characters)  (line-length)
  43:81     error    line too long (85 > 80 characters)  (line-length)
  44:81     error    line too long (94 > 80 characters)  (line-length)
  45:81     error    line too long (81 > 80 characters)  (line-length)
  49:81     error    line too long (83 > 80 characters)  (line-length)
  50:81     error    line too long (81 > 80 characters)  (line-length)
  51:81     error    line too long (83 > 80 characters)  (line-length)
  52:81     error    line too long (87 > 80 characters)  (line-length)
  53:81     error    line too long (81 > 80 characters)  (line-length)
  54:81     error    line too long (82 > 80 characters)  (line-length)
  55:81     error    line too long (89 > 80 characters)  (line-length)
  59:81     error    line too long (87 > 80 characters)  (line-length)
  64:81     error    line too long (91 > 80 characters)  (line-length)
  65:81     error    line too long (83 > 80 characters)  (line-length)
  66:81     error    line too long (88 > 80 characters)  (line-length)
  67:81     error    line too long (81 > 80 characters)  (line-length)
  68:81     error    line too long (84 > 80 characters)  (line-length)
  69:81     error    line too long (82 > 80 characters)  (line-length)
  74:81     error    line too long (88 > 80 characters)  (line-length)
  77:81     error    line too long (82 > 80 characters)  (line-length)
  84:81     error    line too long (82 > 80 characters)  (line-length)
  92:81     error    line too long (88 > 80 characters)  (line-length)
  93:81     error    line too long (85 > 80 characters)  (line-length)
  97:11     error    wrong indentation: expected 12 but found 10  (indentation)
  104:3     error    wrong indentation: expected 4 but found 2  (indentation)

./keycloak.org_keycloakclients_crd.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (86 > 80 characters)  (line-length)
  21:81     error    line too long (81 > 80 characters)  (line-length)
  22:81     error    line too long (152 > 80 characters)  (line-length)
  25:81     error    line too long (82 > 80 characters)  (line-length)
  26:81     error    line too long (82 > 80 characters)  (line-length)
  27:81     error    line too long (156 > 80 characters)  (line-length)
  32:81     error    line too long (85 > 80 characters)  (line-length)
  80:81     error    line too long (81 > 80 characters)  (line-length)
  86:81     error    line too long (84 > 80 characters)  (line-length)
  144:81    error    line too long (84 > 80 characters)  (line-length)
  171:15    error    wrong indentation: expected 16 but found 14  (indentation)
  174:81    error    line too long (82 > 80 characters)  (line-length)
  177:81    error    line too long (89 > 80 characters)  (line-length)
  180:81    error    line too long (89 > 80 characters)  (line-length)
  181:81    error    line too long (85 > 80 characters)  (line-length)
  184:81    error    line too long (83 > 80 characters)  (line-length)
  188:81    error    line too long (82 > 80 characters)  (line-length)
  189:81    error    line too long (82 > 80 characters)  (line-length)
  193:81    error    line too long (89 > 80 characters)  (line-length)
  194:81    error    line too long (84 > 80 characters)  (line-length)
  195:81    error    line too long (83 > 80 characters)  (line-length)
  196:81    error    line too long (83 > 80 characters)  (line-length)
  202:21    error    wrong indentation: expected 22 but found 20  (indentation)
  209:81    error    line too long (82 > 80 characters)  (line-length)
  210:81    error    line too long (82 > 80 characters)  (line-length)
  211:81    error    line too long (82 > 80 characters)  (line-length)
  212:81    error    line too long (86 > 80 characters)  (line-length)
  217:11    error    wrong indentation: expected 12 but found 10  (indentation)
  221:81    error    line too long (88 > 80 characters)  (line-length)
  224:81    error    line too long (82 > 80 characters)  (line-length)
  231:81    error    line too long (82 > 80 characters)  (line-length)
  239:81    error    line too long (88 > 80 characters)  (line-length)
  240:81    error    line too long (85 > 80 characters)  (line-length)
  244:11    error    wrong indentation: expected 12 but found 10  (indentation)
  251:3     error    wrong indentation: expected 4 but found 2  (indentation)

./keycloak.org_keycloaks_crd.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (86 > 80 characters)  (line-length)
  21:81     error    line too long (81 > 80 characters)  (line-length)
  22:81     error    line too long (152 > 80 characters)  (line-length)
  25:81     error    line too long (82 > 80 characters)  (line-length)
  26:81     error    line too long (82 > 80 characters)  (line-length)
  27:81     error    line too long (156 > 80 characters)  (line-length)
  35:81     error    line too long (81 > 80 characters)  (line-length)
  44:81     error    line too long (82 > 80 characters)  (line-length)
  49:81     error    line too long (82 > 80 characters)  (line-length)
  50:81     error    line too long (83 > 80 characters)  (line-length)
  51:81     error    line too long (91 > 80 characters)  (line-length)
  52:81     error    line too long (82 > 80 characters)  (line-length)
  53:81     error    line too long (87 > 80 characters)  (line-length)
  54:81     error    line too long (85 > 80 characters)  (line-length)
  55:81     error    line too long (88 > 80 characters)  (line-length)
  56:81     error    line too long (82 > 80 characters)  (line-length)
  57:81     error    line too long (107 > 80 characters)  (line-length)
  58:81     error    line too long (83 > 80 characters)  (line-length)
  59:81     error    line too long (86 > 80 characters)  (line-length)
  60:81     error    line too long (81 > 80 characters)  (line-length)
  61:81     error    line too long (88 > 80 characters)  (line-length)
  62:81     error    line too long (88 > 80 characters)  (line-length)
  63:81     error    line too long (91 > 80 characters)  (line-length)
  67:81     error    line too long (90 > 80 characters)  (line-length)
  72:81     error    line too long (81 > 80 characters)  (line-length)
  75:81     error    line too long (82 > 80 characters)  (line-length)
  83:81     error    line too long (81 > 80 characters)  (line-length)
  86:81     error    line too long (84 > 80 characters)  (line-length)
  90:81     error    line too long (82 > 80 characters)  (line-length)
  97:81     error    line too long (82 > 80 characters)  (line-length)
  105:81    error    line too long (88 > 80 characters)  (line-length)
  106:81    error    line too long (85 > 80 characters)  (line-length)
  113:11    error    wrong indentation: expected 12 but found 10  (indentation)
  123:3     error    wrong indentation: expected 4 but found 2  (indentation)
```

**Comment by abstractj** [comment]
@slaskawi overall LGTM, but I would like to hear your opinion about the YAML issues. Maybe, worth to fix them. Great stuff btw.

**Comment by slaskawi** [comment]
@abstractj Those are auto-generated files. A typical workflow includes:
- Modifying Go types
- Running `operator-sdk generate openapi`

I think using longer form of field description (the one with `|` symbol) or starting with `----` should be Operator SDK responsibility. Perhaps it's worth to file them a JIRA?

**Pull request merged**